partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rel/RelOpWindow.kt

```diff
@@ -0,0 +1,165 @@
+package org.partiql.eval.internal.operator.rel
+
+import org.partiql.eval.Environment
+import org.partiql.eval.ExprRelation
+import org.partiql.eval.ExprValue
+import org.partiql.eval.Row
+import org.partiql.eval.WindowFunction
+import org.partiql.eval.WindowPartition
+import org.partiql.eval.internal.helpers.DatumArrayComparator
+
+/**
+ * Assume input has been sorted.
+ */
+internal class RelOpWindow(
+    private val input: ExprRelation,
+    private val functions: List<WindowFunction>,
+    private val partitionBy: List<ExprValue>,
+    private val sortBy: List<Collation>
+) : RelOpPeeking() {
+
+    private companion object {
+        private val comparator = DatumArrayComparator
+    }
+
+    private lateinit var _env: Environment
+
+    /**
+     * This is used to track the current partition that we are iterating through.
+     */
+    private var _partition: LocalPartition = LocalPartition()
+
+    /**
+     * This is used to track where we are in the partition that we are currently iterating through.
+     * @see peek
+     */
+    private var _partitionPeekingNumber: Long = 0
+
+    /**
+     * When lazily creating the partition, we need to step out-of-bounds from the current partition to know
+     * whether we are in a new partition. So, we need to put the out-of-bounds row into a place to be used
+     * when creating the next partition.
+     */
+    private var leftoverRow: Row? = null
+
+    override fun openPeeking(env: Environment) {
+        input.open(env)
+        this._env = env
+        _partitionPeekingNumber = -1L
+        _partition = LocalPartition()
+        functions.map { it.reset(_partition) }
+    }
+
+    override fun peek(): Row? {
+        // Check if there is an existing partition. If so, evaluate and return.
+        _partitionPeekingNumber++
+        if (_partition.size() > _partitionPeekingNumber) {
+            return produceResult()
+        }
+        _partitionPeekingNumber = 0L
```

Comment:
Notes: When we reached here, it means we either at the very first peek call, or we drained the previous partition.